### PR TITLE
[Backend] Creating a poll object

### LINKED ISSUE #5315: Polls implementation: [Backend] Creating a poll object
## Issue - Enhancement
This issue relates to the backend implementation of creating a poll. Related frontend issue #5146 

### Acceptance Criteria (optional)
- [ ] Create a `CreatePoll` mutation
- [ ] Poll is correctly saved to database
- [ ] Poll options are correctly saved to database

**Estimated effort:** 24h

### PR DESCRIPTION
**Pull request by BartoszJarocki**
Fixes #5315. This PR depends on #5303. 

Scope
* Added `CreatePoll` mutation
* Added `getPollsById`, `getPollsByDiscussionId` and `getPollOptionsByPollId` db queries
* Added `pollLoaders` 
* Added GraphQL types related to polls
* Updated `pgtyped` to the newest version to support readonly arrays (no more nasty casts when passing parameters to generated queries!) 

### Acceptance Criteria (optional)
- [ ] Create a `CreatePoll` mutation
- [ ] Poll is correctly saved to database
- [ ] Poll options are correctly saved to database

Example mutation in graphiql (change `teamId` and `discussionId` to the proper values that exists in your local db)
```
mutation {
  createPoll(
    newPoll: {threadSortOrder: 0.4, teamId: "6gEbA4BOk9", discussionId: "6j6SB4HMXK", title: "poll!", options: [{title: "option title1"}, {title: "option title2"}]}
  ) {
    ... on ErrorPayload {
      error {
        title
        message
      }
    }
    ... on CreatePollSuccess {
      poll {
        id
        title
        discussionId
        threadSortOrder
        options {
          id
          title
        }
      }
    }
  }
}
```

Example result
Screenshot 2021-09-01 at 07 25 00

### CODE REVIEW

File: packages/client/mutations/CreatePollMutation.ts
Comment (BartoszJarocki) [inline packages/client/mutations/CreatePollMutation.ts:38]: This will be implemented in frontend part #5146

**Comment by BartoszJarocki** [comment]
@tianrunhe may I ask you to take a look, please? 🙏🏻

File: packages/server/postgres/queries/src/getPollsOptionsByPollIdQuery.sql
Comment (tianrunhe) [inline packages/server/postgres/queries/src/getPollsOptionsByPollIdQuery.sql:2]: nit: make the name consistent with file name

File: packages/server/postgres/queries/getPollOptionsByPollId.ts
Comment (tianrunhe) [inline packages/server/postgres/queries/getPollOptionsByPollId.ts:1]: Nit: rename this file to `getPollOptionsByPollId`**`s`**

File: packages/server/postgres/queries/src/getPollsByDiscussionIdQuery.sql
Comment (tianrunhe) [inline packages/server/postgres/queries/src/getPollsByDiscussionIdQuery.sql:2]: Nit: Rename it to `getPollsByDiscussionId`**`s`**`Query`.

File: packages/server/postgres/queries/getPollsByDiscussionId.ts
Comment (tianrunhe) [inline packages/server/postgres/queries/getPollsByDiscussionId.ts:1]: Rename this to `getPollsByDiscussionId`**`s`** so it's clear that it expects an array of discussionIds

File: packages/server/postgres/queries/getPollsByDiscussionId.ts
Comment (tianrunhe) [inline packages/server/postgres/queries/getPollsByDiscussionId.ts:4]: Nit: try to make this consistent with the file name

File: packages/server/postgres/queries/src/getPollsByIdQuery.sql
Comment (tianrunhe) [inline packages/server/postgres/queries/src/getPollsByIdQuery.sql:2]: Rename this to `getPollsById`**`s`**`Query`

File: packages/server/postgres/queries/getPollsById.ts
Comment (tianrunhe) [inline packages/server/postgres/queries/getPollsById.ts:1]: Rename this file to `getPollsbyId`**`s`** so it's clear that it expects an array of ids

File: packages/server/postgres/queries/getMeetingTaskEstimates.ts
Comment (tianrunhe) [inline packages/server/postgres/queries/getMeetingTaskEstimates.ts:12]: Is this change of removing `meetingIds` from the argument list expected?

File: packages/server/dataloader/pollsLoaders.ts
Comment (tianrunhe) [inline packages/server/dataloader/pollsLoaders.ts:13]: I think if the query returns nothing for the `pollIds`, it will return an empty array so this feels redundant

File: packages/server/graphql/types/Poll.ts
Comment (tianrunhe) [inline packages/server/graphql/types/Poll.ts:25]: "the **poll** was created in"

File: packages/server/graphql/types/Poll.ts
Comment (tianrunhe) [inline packages/server/graphql/types/Poll.ts:44]: Clear copy/paste typo 🤣

File: packages/server/graphql/mutations/createPoll.ts
Comment (tianrunhe) [inline packages/server/graphql/mutations/createPoll.ts:115]: `meetingId` is missing here but it's included in `CreatePollPayload`

File: packages/server/graphql/mutations/createPoll.ts
Comment (tianrunhe) [inline packages/server/graphql/mutations/createPoll.ts:90]: I'm curious why you chose to not include `meetingId` as another input argument?

File: packages/server/graphql/mutations/createPoll.ts
Comment (tianrunhe) [inline packages/server/graphql/mutations/createPoll.ts:56]: Should we also check if the `title` is a duplicate, i.e., can we create two identical polls?

**Review by tianrunhe — CHANGES_REQUESTED** [review]

File: packages/server/postgres/queries/getMeetingTaskEstimates.ts
Comment (BartoszJarocki) [reply packages/server/postgres/queries/getMeetingTaskEstimates.ts:12]: Yes, if you take a look at the underlying query it doesn't have `meetingIds` and after upgrading `pgtyped` TS complained
```
 Object literal may only specify known properties, and 'meetingIds' does not exist in type 'IGetLatestTaskEstimatesQueryParams'.ts(2345)
```

File: packages/server/graphql/types/Poll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/types/Poll.ts:25]: 🤦🏻‍♂️

File: packages/server/graphql/types/Poll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/types/Poll.ts:44]: Inheritance by copy/paste has some downsides ;) I should've noticed that during self-review 🙏🏻

File: packages/server/graphql/mutations/createPoll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/mutations/createPoll.ts:115]: oh, good catch! I don't think it should be included there so I'm going to remove `meetingId` from `CreatePollPayload`

File: packages/server/graphql/mutations/createPoll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/mutations/createPoll.ts:90]: Including `meetingId` would require another check. We can only create a poll in a meeting related to this particular discussion. So I'd say, _not_ having `meetingId` in input argument actually makes it harder to provide wrong arguments.

File: packages/server/graphql/mutations/createPoll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/mutations/createPoll.ts:56]: I'm personally don't think it should be an issue but I'm open to check that. @mattkrick @enriquesanchez any thoughts on that?

**Comment by BartoszJarocki** [comment]
@tianrunhe thank you for the review! May I ask you to do another round? 🙏🏻

**Review by tianrunhe — APPROVED** [review]
LGTM!

File: packages/server/graphql/mutations/createPoll.ts
Comment (mattkrick) [reply packages/server/graphql/mutations/createPoll.ts:56]: that's a good consideration.
I feel like the `sortOrder` acts as the unique identifier in the UI. 
If they have multiple rounds of polls, removing 1 option each time, they might keep the same name? IMO that's kinda lame, but i could see folks doing it. I'm cool with leaving it out for now just to get it out & see what folks do with it

Comment (mattkrick) [reply packages/server/graphql/mutations/createPoll.ts:56]: can delete this conditional or change it to `title.length === 0` because title is guaranteed to be non-null from `CreatePollInput`

maybe we add a min length?

File: packages/server/graphql/mutations/createPoll.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/createPoll.ts:16]: min title length? maybe 2?

File: packages/server/graphql/mutations/createPoll.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/createPoll.ts:14]: why 4? In other places we use 12 to signify "a few" (1 for each month). maybe that's too much here, but i worry 4 is too little. @enriquesanchez thoughts?

File: packages/server/graphql/mutations/createPoll.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/createPoll.ts:84]: nit: `option.title.length === 0` just to be super explicit (since non-null is guaranteed)

File: packages/server/graphql/mutations/createPoll.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/createPoll.ts:24]: remove `teamId` here. `teamId` lives on the `Discussion` object provided by `discussionId`

File: packages/server/graphql/mutations/createPoll.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/createPoll.ts:51]: can remove, we'll check `Discussion` for this

File: packages/server/graphql/mutations/createPoll.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/createPoll.ts:91]: let's remove this check. in the future, we might make `Discussion.meetingId` nullable.
Instead, let's check `teamId` and make sure they're a team member

File: packages/server/graphql/mutations/createPoll.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/createPoll.ts:101]: this is our first multi-table insert in PG :tada: 
we could do this a few ways:
- chained inserted as you have
- chained inserts in a transaction
- CTEs: https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-MODIFYING

which do you prefer? any pros/cons of each?

one thing worth considering: if the client calls `createPoll` 2 times at once (maybe their internet dropped & reconnected, or we goof up the front-end & let them click submit twice, etc) is there a way we can de-dupe the poll creation?

File: packages/server/graphql/types/CreatePollInput.ts
Comment (mattkrick) [inline packages/server/graphql/types/CreatePollInput.ts:14]: remove teamId

File: packages/server/graphql/types/CreatePollInput.ts
Comment (mattkrick) [inline packages/server/graphql/types/CreatePollInput.ts:24]: polls don't have a `threadParentId` because they can't be a "reply" to another type, right?

File: packages/server/graphql/types/CreatePollPayload.ts
Comment (mattkrick) [inline packages/server/graphql/types/CreatePollPayload.ts:9]: nit: sometimes i like to include the field `pollId` in case a future use case only needs the Id instead of fetching the whole poll

File: packages/client/shared/gqlIds/PollId.ts
Comment (mattkrick) [inline packages/client/shared/gqlIds/PollId.ts:2]: nit: the other files in the folder have a `join` method. I know where' not actually joining anything here though. not sure which is better :shrug:

File: packages/server/graphql/types/PollOption.ts
Comment (mattkrick) [inline packages/server/graphql/types/PollOption.ts:29]: type is Team?

File: packages/server/graphql/types/Poll.ts
Comment (mattkrick) [inline packages/server/graphql/types/Poll.ts:24]: `meetingId` is nullable in the DB

File: packages/server/graphql/types/PollOption.ts
Comment (mattkrick) [inline packages/server/graphql/types/PollOption.ts:25]: this is where stuff gets dicey.
is this the numerical database ID or the GraphQL ID? e.g.  `1` or `PollOption:1`?
we'll have to be very explicit on which one it is in the description & make sure we always keep it consistent.
My advice is:
- any DB query helper should return the DB foreign key
- any dataloader should return the DB foreign key
- any graphql resolver should return the GraphQL ID

this will keep our database ID abstraction from leaking into the graphql schema. 
 that way, on our client cache, we'll still have globally unique IDs so we can do this: https://relay.dev/docs/guided-tour/reusing-cached-data/filling-in-missing-data/

File: packages/server/graphql/types/PollOption.ts
Comment (mattkrick) [inline packages/server/graphql/types/PollOption.ts:35]: i foresee a `voteUsers` array too, but we can hold off until the front-end gets built

File: packages/server/graphql/types/Threadable.ts
Comment (mattkrick) [inline packages/server/graphql/types/Threadable.ts:71]: might be worthwhile casting to the `database/Threadable.ts` type so we get a little type safety here. 
then again, you could just merge that file into this one

File: packages/server/postgres/queries/src/getPollOptionsByPollIdsQuery.sql
Comment (mattkrick) [inline packages/server/postgres/queries/src/getPollOptionsByPollIdsQuery.sql:5]: I keep wondering if we'll ever query a PollOption without querying a Poll, or vice versa.
In NoSQL, we'd simply have all the options live in the same document as the poll.
In SQL, it makes sense to properly normalize them. 

Is it worthwhile writing 1 query that fetches a poll and all its options? that means we could get everything we need with a single trip to the database.

**Review by mattkrick — CHANGES_REQUESTED** [review]
great work! all the comments are pretty tiny, feel free to discuss

File: packages/server/dataloader/customLoaderMakers.ts
Comment (BartoszJarocki) [reply packages/server/dataloader/customLoaderMakers.ts:155]: sure!

File: packages/server/postgres/queries/getMeetingTaskEstimates.ts
Comment (BartoszJarocki) [reply packages/server/postgres/queries/getMeetingTaskEstimates.ts:13]: makes sense 😅  fixed!

File: packages/server/dataloader/RethinkDataLoader.ts
Comment (BartoszJarocki) [reply packages/server/dataloader/RethinkDataLoader.ts:22]: @mattkrick which one do you prefer? I think it's better to separate loaders by feature as it's easier to find/maintain.
Also, by extending the type you mean this
```
type CustomLoaderMakers = typeof customLoaderMakers & typeof atlassianLoaders & typeof pollLoaders
```

ie. adding `pollLoaders` here? 

And thank you, I missed that!

File: packages/server/graphql/mutations/createPoll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/mutations/createPoll.ts:14]: I used 4 as this is defined in [AC](https://github.com/ParabolInc/parabol/issues/5146). We can change that to whatever we think is best, thought!

File: packages/server/graphql/mutations/createPoll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/mutations/createPoll.ts:16]: Good idea! I'll move all the consts related to polls to `constEnums` and clean the code.

File: packages/server/graphql/mutations/createPoll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/mutations/createPoll.ts:51]: Hmm, not sure I understand. Should we remove this check entirely or read `teamId` from the discussion and then perform the check? If we're going to remove this entirely, how can we ensure that user with a given access token has the permission to add polls to a given discussion? 🤔

File: packages/server/graphql/types/CreatePollInput.ts
Comment (BartoszJarocki) [reply packages/server/graphql/types/CreatePollInput.ts:24]: Oh yeah. The comment should not say `relative to threadParentId`. Just to make sure, `threadSortOrder` is still a valid property used to determine the sort order in the discussion, or am I thinking wrong here? 🙏

File: packages/server/graphql/types/PollOption.ts
Comment (BartoszJarocki) [reply packages/server/graphql/types/PollOption.ts:29]: Not sure how did I miss that, sorry!🤦

File: packages/client/shared/gqlIds/PollId.ts
Comment (BartoszJarocki) [reply packages/client/shared/gqlIds/PollId.ts:2]: Join it is!

File: packages/server/graphql/types/Poll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/types/Poll.ts:24]: Good catch!

File: packages/server/graphql/types/CreatePollPayload.ts
Comment (BartoszJarocki) [reply packages/server/graphql/types/CreatePollPayload.ts:9]: Just to make sure, with a format of `poll:idGeneratedByDatabase`, not raw id generated by DB? 😅

Comment (BartoszJarocki) [reply packages/server/graphql/types/CreatePollPayload.ts:9]: This was a dumb question, it's a resolver so can't be anything else but `poll:idGeneratedByDatabase` to keep things consistent.

File: packages/server/graphql/mutations/createPoll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/mutations/createPoll.ts]: I guess the answer is reading `teamId` from the discussion, correct me if I'm wrong, please!

File: packages/server/graphql/types/Threadable.ts
Comment (BartoszJarocki) [reply packages/server/graphql/types/Threadable.ts:71]: @mattkrick I've decided to leave `database/Threadable.ts` in a separate file as there's more than one place that uses this type and in my opinion, it's cleaner that way. I've added type checks, though. Let me know what do you think!

File: packages/server/graphql/types/PollOption.ts
Comment (BartoszJarocki) [reply packages/server/graphql/types/PollOption.ts:25]: This is a very good catch! Added a better comment and a fix to provide a proper id. 🙏

File: packages/server/graphql/mutations/createPoll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/mutations/createPoll.ts]: Implemented with CTE, thank you for pointing that out!

File: packages/server/graphql/types/Threadable.ts
Comment (BartoszJarocki) [inline packages/server/graphql/types/Threadable.ts:72]: I'm not sure if `ThreadableDB` is the best name but there's nothing else that came to my mind. If you there's a better name to avoid collision with gql `Threadable` I'm happy to improve~

**Comment by BartoszJarocki** [comment]
@mattkrick I'd appreciate it if you could take a look. I've fixed all the feedback. One thing is still left, related to fetching polls and poll options at once but it's not a big deal in terms of the performance so I'd leave it as-is for now. Thank you!

File: packages/server/graphql/types/Poll.ts
Comment (mattkrick) [inline packages/server/graphql/types/Poll.ts:52]: what are your plans with the connection? unless we're going to return a list of 20+ polls to the viewer we can probably remove poll connection & edge

File: packages/server/graphql/types/Threadable.ts
Comment (mattkrick) [reply packages/server/graphql/types/Threadable.ts:72]: i'm right there with ya, i've seen some people call it `CThreadable` where `C` is for `class`, but i think the DB suffix is a little more expressive

File: packages/server/postgres/queries/src/insertPollWithOptionsQuery.sql
Comment (mattkrick) [inline packages/server/postgres/queries/src/insertPollWithOptionsQuery.sql:16]: this is the line that caused all the headaches i bet!

**Comment by mattkrick** [comment]
LGTM, only question outstanding is if we need the PollConnection or not. Good to merge after you decide!

File: packages/server/postgres/queries/src/insertPollWithOptionsQuery.sql
Comment (BartoszJarocki) [reply packages/server/postgres/queries/src/insertPollWithOptionsQuery.sql:16]: Oh yeah. One thing I've learned about writing SQL is that you don't know the solution, start with a super simple query, data and then go towards the end goal. SQL errors are basically useless so it's hard to find out what's wrong with the query.

File: packages/server/graphql/types/Poll.ts
Comment (BartoszJarocki) [reply packages/server/graphql/types/Poll.ts:52]: You're right, we don't need that. My initial thought was that every threadable would have to expose the connection but now I know it's not true. Removing! ✂️

**Pull request merged**
